File: src/examples/eval/src/app/csa104Comp/msgFns.c
#include <assert.h>
#include "msgFns.h"

SaMsgHandleT            msgLibraryHandle;
SaMsgQueueHandleT       msgQueueHandle;

void msgInitialize(void)
{
    SaAisErrorT             Rc;
    SaMsgCallbacksT         MsgCallbacks = {
                                .saMsgQueueOpenCallback           = (SaMsgQueueOpenCallbackT) 0,
                                .saMsgQueueGroupTrackCallback     = (SaMsgQueueGroupTrackCallbackT) 0,
                                .saMsgMessageDeliveredCallback    = (SaMsgMessageDeliveredCallbackT) 0,
                                .saMsgMessageReceivedCallback     = (SaMsgMessageReceivedCallbackT) 0,
                            };
    SaVersionT              Version = {
                                .majorVersion = 1,
                                .minorVersion = 1,
                                .releaseCode  = 'B',
                            };

	Rc = saMsgInitialize (& msgLibraryHandle, & MsgCallbacks, & Version);
	if ( SA_AIS_OK != Rc)
    {
		clprintf ( CL_LOG_SEV_ERROR, "Init failed [0x%X]", Rc);
        assert(0);
    }
}


SaAisErrorT msgOpen(const char* queuename,int bytesPerPriority)
{
	SaAisErrorT             rc;
    SaNameT                 saQueueName;
    SaMsgQueueCreationAttributesT   CreationAttributes;
    
    saQueueName.length=strlen(queuename);
    memcpy(saQueueName.value,queuename,saQueueName.length+1);

    SaMsgQueueOpenFlagsT    OpenFlags = SA_MSG_QUEUE_CREATE;
    
    CreationAttributes.creationFlags = 0;
    for (int i=0;i<SA_MSG_MESSAGE_LOWEST_PRIORITY;i++)
        CreationAttributes.size[i] = bytesPerPriority;
    CreationAttributes.retentionTime = 0;

    rc = saMsgQueueOpen (msgLibraryHandle, &saQueueName, & CreationAttributes, OpenFlags, SA_TIME_MAX, &msgQueueHandle);
    if (SA_AIS_OK != rc)
    {
		clprintf ( CL_LOG_SEV_ERROR, "Msg QueueOpen failed [0x%X]\n\r", rc);
    }
    return rc;
 }

SaAisErrorT msgSend(const char* queuename, void* buffer, int length)
{
    SaAisErrorT             rc;
    SaNameT                 saQueueName;
    SaMsgMessageT           message;

    /* Load the SAF string */
    saQueueName.length=strlen(queuename);
    memcpy(saQueueName.value,queuename,saQueueName.length+1);

    /* Load the SAF message structure */
    message.type = 0;
    message.version.releaseCode = 0;
    message.version.majorVersion=0;
    message.version.minorVersion=0;
    message.senderName = 0;  /* You could put a SaNameT* in here if you wanted to pass a reply queue (for example) */
    message.size = length;
    message.data = buffer;
    message.priority = SA_MSG_MESSAGE_HIGHEST_PRIORITY;
    rc = saMsgMessageSend (msgLibraryHandle, &saQueueName, &message, SA_TIME_MAX);
    if (SA_AIS_OK != rc)
    {
        /* Error 0xC here means that the queue has not yet been created.
         That is, the receiver is not yet listening. */
		clprintf ( CL_LOG_SEV_ERROR, "Msg saMsgMessageSend to queue [%s] failed [0x%X]", saQueueName.value,rc );
    }
    return rc;
}


void* msgReceiverLoop(void * notused)
{
    SaAisErrorT             rc;
    SaNameT                 SenderName;
    char                    Data[1024];
    SaMsgSenderIdT          SenderId;
    SaTimeT                 SendTime;
    
    while(1)
    {
        
        SaMsgMessageT           message = {
            .size       = 1024,
            .senderName = &SenderName,
            .data       = Data,
        };
    
        rc = saMsgMessageGet (msgQueueHandle, &message, & SendTime, & SenderId, SA_TIME_MAX);
        if (SA_AIS_OK != rc)
        {
            clprintf ( CL_LOG_SEV_ERROR, "Msg saMsgMessageGet failed [0x%X]\n\r", rc );
            break;
        }
        
        if (message.senderName->length)
        {
            clprintf ( CL_LOG_SEV_INFO, "Sender Name   : %s\n", message.senderName->value);
        }
        clprintf ( CL_LOG_SEV_INFO, "Received Message  : %s\n", (char *)message.data);
        
    }
    
    rc = saMsgQueueClose (msgQueueHandle);
    if (SA_AIS_OK != rc)
    {
		clprintf ( CL_LOG_SEV_ERROR, "Msg Queue Close failed [0x%X]\n\r", rc );
    }
    return 0;
 
}

File: src/examples/java/msg/msgUtils/MsgFns.java
package msgUtils;

import com.sun.jna.Pointer;
import saMsg.SaMsgCallbacksT;
import saAis.SaAisLibrary;
import saAis.SaAisLibrary.SaVersionT;
import saAis.SaAisLibrary.SaNameT;
import saAis.SaAisLibrary.SaAisErrorT;
import saMsg.SaMsgQueueCreationAttributesT;
import java.nio.charset.StandardCharsets;
import saMsg.SaMsgClientLibrary;
import com.sun.jna.ptr.LongByReference;
import saMsg.SaMsgClientLibrary.SaMsgQueueGroupChangesT;
import msgReceiver.ClReceiverMain;

public class MsgFns {
  //private static final String ACTIVE_COMP_QUEUE = "csa104msgqueue";
  public static final int QUEUE_LENGTH = 2048;
  public static final String[] queues = {"test1"};//,"test2", "abcd", "efghigke"};
  public static saMsg.SaMsgClientLibrary saMsgLib = saMsg.SaMsgClientLibrary.INSTANCE;
  private static LongByReference msgLibraryHandle = new LongByReference();
  public static LongByReference[] msgQueueHandle = new LongByReference[queues.length];
  public static int numQueues = 0;
  
  public static int msgInitialize() {
    for (int i = 0;i<msgQueueHandle.length;i++) {
      msgQueueHandle[i] = new LongByReference();
    }
    int rc;
    saMsg.SaMsgCallbacksT msgCallbacks = new saMsg.SaMsgCallbacksT(null,null,null,null);
    msgCallbacks.write();
    saAis.SaAisLibrary.SaVersionT ver = new saAis.SaAisLibrary.SaVersionT();
    ver.releaseCode = 'B';
    ver.majorVersion = 01;
    ver.minorVersion = 01;
    ver.write();
	  rc = saMsgLib.saMsgInitialize (msgLibraryHandle, msgCallbacks, ver.getPointer());
	  if (saAis.SaAisLibrary.SaAisErrorT.SA_AIS_OK != rc) {
		  assert false : String.format("Msg init failed [0x%X]", rc);      
    }
    return rc;
  }
  public static int msgOpen(String queueName, int bytesPerPriority) {
    int rc = saAis.SaAisLibrary.SaAisErrorT.SA_AIS_OK;
    //byte[] bytes = (queueName + "\0").getBytes(StandardCharsets.UTF_8);
    //byte[] bytes = queueName.getBytes(StandardCharsets.UTF_8);
    //saAis.SaAisLibrary.SaNameT saQueueName = new saAis.SaAisLibrary.SaNameT((short)(bytes.length), bytes);
    //saQueueName.write();
    
    byte[] bytes = queueName.getBytes(StandardCharsets.UTF_8);
    saAis.SaAisLibrary.SaNameT saQueueName = new saAis.SaAisLibrary.SaNameT();
    saQueueName.clear();
    saQueueName.length = (short) bytes.length;
    System.arraycopy(bytes, 0, saQueueName.value, 0, bytes.length);
    saQueueName.write();    
    
    saMsg.SaMsgQueueCreationAttributesT creationAttributes = new saMsg.SaMsgQueueCreationAttributesT();    
    
    int openFlags = saMsg.SaMsgClientLibrary.SA_MSG_QUEUE_CREATE;
    
    creationAttributes.creationFlags = 0;
    int i;
    for (i=0;i<saMsg.SaMsgClientLibrary.SA_MSG_MESSAGE_LOWEST_PRIORITY;i++) {
      creationAttributes.size[i] = bytesPerPriority;
    }
    creationAttributes.retentionTime = 0;
    creationAttributes.write();
    //ClReceiverMain.clprintf(clUtils.ClUtilsLibrary.ClLogSeverityT.CL_LOG_SEV_INFO, String.format("csa104: Opening msg queue name [%s] length [%d]", saQueueName.toNameString(), saQueueName.length));
    rc = saMsgLib.saMsgQueueOpen(msgLibraryHandle.getValue(), saQueueName, creationAttributes, openFlags, saAis.SaAisLibrary.SA_TIME_END, msgQueueHandle[numQueues]);    
    numQueues++;
    return rc;
  }
  public static int msgSend(String queueName, Pointer buffer, int length) {
    int rc = saAis.SaAisLibrary.SaAisErrorT.SA_AIS_OK;
    /*byte[] bytes = (queueName + "\0").getBytes(StandardCharsets.UTF_8);
    saAis.SaAisLibrary.SaNameT saQueueName = new saAis.SaAisLibrary.SaNameT((short)(queueName.length()+1), bytes);
    saQueueName.write();*/
    
    byte[] bytes = queueName.getBytes(StandardCharsets.UTF_8);
    saAis.SaAisLibrary.SaNameT saQueueName = new saAis.SaAisLibrary.SaNameT();
    saQueueName.clear();
    saQueueName.length = (short) bytes.length;
    System.arraycopy(bytes, 0, saQueueName.value, 0, bytes.length);
    saQueueName.write(); 
    
    saMsg.SaMsgMessageT message = new saMsg.SaMsgMessageT();   

    /* Load the SAF message structure */
    message.type = 0;
    saAis.SaAisLibrary.SaVersionT version = new saAis.SaAisLibrary.SaVersionT();
    version.releaseCode = 0;
    version.majorVersion=0;
    version.minorVersion=0;
    version.write();
    message.version = version;
    message.senderName = null;  /* You could put a SaNameT* in here if you wanted to pass a reply queue (for example) */
    message.size = length;
    message.data = buffer;
    message.priority = saMsg.SaMsgClientLibrary.SA_MSG_MESSAGE_HIGHEST_PRIORITY;
    message.write();
    rc = saMsgLib.saMsgMessageSend (msgLibraryHandle.getValue(), saQueueName.getPointer(), message, saAis.SaAisLibrary.SA_TIME_END);
    
    return rc;
  }
  /*public void msgReceiverLoop() {
    
  }*/
}
